### PR TITLE
Clusters Client

### PR DESCRIPTION
**Pull request by luizbafilho**
Closes: https://github.com/weaveworks/weave-gitops/issues/1777

**What changed?**
Introduces the Clusters Client, this is an abstraction to help make calls to all the clusters. It is a thin wrapper around the `controller-runtime/client` that handles the connections to the clusters and when contacting them does it in parallel.

**Why?**
- Simplify resources updates and listing

**How did you test it?**
- automated and manual tests

**Release notes**

**Documentation Changes**

### CODE REVIEW

**Review by jpellizzari — APPROVED** [review]
LGTM

File: core/clustersmngr/client.go [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	ErrClusterNotFound error = errors.New("cluster not found")
```
Comment (Callisto13) [inline core/clustersmngr/client.go:13]: can we turn this into an error type that lets us pass in the cluster name? might be useful for debugging

File: core/logger/logger_test.go [hunk 2]
```diff
@@ -157,7 +157,7 @@ func TestBuildConfigWithOptions(t *testing.T) {
 }
 
 var (
-	rfc3339Regexp = `^((\d{4}-\d{2}-\d{2})T(\d{2}:\d{2}:\d{2}(?:\.\d+)?))(Z|[\+-]\d{2}:\d{2})$`
+	rfc3339Regexp = `^((\d{4}-\d{2}-\d{2})T(\d{2}:\d{2}:\d{2}(?:\.\d+)?))(Z|[\+-]\d{2}:?\d{2})$`
```
Comment (Callisto13) [inline core/logger/logger_test.go:160]: i think balazs fixed this in his last pr, check on main?

**Review by Callisto13 — COMMENTED** [review]
like this a lot! great work! just a couple of questions/requests

File: core/clustersmngr/client.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+				errs = append(errs, fmt.Errorf("cluster=\"%s\" err=\"%s\"", clusterName, err))
+			}
+		}(clusterName, c)
+	}
+
+	wg.Wait()
+
+	if len(errs) > 0 {
+		return fmt.Errorf("failed to list resources: %s", errs)
+	}
```
[the remark this answers is not in the document]
Comment (luizbafilho) [reply core/clustersmngr/client.go:94]: I made myself the same question, and I don't think the end goal would be this, but we need to coordinate with the UI to settle down the best way to handle this "partial success" response.

File: core/logger/logger_test.go (hunk 2 above)
Comment (luizbafilho) [reply core/logger/logger_test.go]: just pulled from master

File: core/clustersmngr/client.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+		pool: clientsPool,
+	}
+}
+
+func (c *clustersClient) ClientsPool() ClientsPool {
+	return c.pool
+}
+
+func (c *clustersClient) Get(ctx context.Context, cluster string, key client.ObjectKey, obj client.Object) error {
+	client := c.pool.Clients()[cluster]
```
Comment (yitsushi) [inline core/clustersmngr/client.go:57]: that `c.pool.Clients()[cluster]` hurts my eyes a bit, totally personal taste. I know it does not fail, but my brain constantly says "it should fail if the key is not defined".

```go
func (cp *clientsPool) Client(name string) (client.Client, error) {
  if c, found := cp.clients[name]; found && c != nil {
    return c, nil
  }

  return ClusterNotFoundError{Cluster: name}
}

client, err := c.pool.Client(cluster)
if err != nil {
  return err
}
```

Not a blocker at all. That's only an extra sugar.

Comment (luizbafilho) [reply core/clustersmngr/client.go:57]: agree, much better, pushing the change

**Review by Callisto13 — APPROVED** [review]

**Pull request merged**